HIP/hip-792.md

```diff
@@ -0,0 +1,290 @@
+and therefore a smart contract may not provide its authorization of a transaction
+by any means other than its own execution when invoked within a transaction.
+In other words smart contract authorization must necessarily occur while a transaction is "in flight",
+and therefore cannot occur client-side before being submitted to the network.
+
+However the latter scenario with smart contract IDs
+within "complex" keys **should** be supported, as noted in
+[Hedera's documentation on Keys](https://docs.hedera.com/hedera/sdks-and-apis/hedera-api/basic-types/key):
+
+> Note that when a Key is a smart contract ID,
+> it doesn't mean the contract with that ID will actually create a cryptographic signature.
+> It only means that when the contract calls a precompiled contract,
+> the resulting "child transaction" will be authorized to perform any action controlled by the Key.
+
+Therefore this HIP proposes a means, via the `hederaAccountSystem` system contract,
+to verify authorization of a transaction that will work in the above scenarios.
+
+> NOTE:
+>
+> Hedera Token Service already exposes a system contract `hederaTokenService`,
+> and its function already performs the same authorization as proposed
+> for `hederaAccountService.isAuthorizedCurrentTransaction()`
+>
+> This works as system contracts, including `hederaTokenService`,
+> are passed in a [`TxnAwareEvmSigsVerifier`](https://github.com/hashgraph/hedera-services/blob/810971de2b/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/sources/TxnAwareEvmSigsVerifier.java#L60),
+> which has a [`TransactionContext`](https://github.com/hashgraph/hedera-services/blob/810971de2b/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/sources/TxnAwareEvmSigsVerifier.java#L70)
+> that contains the necessary information in [`JKey`](https://github.com/hashgraph/hedera-services/blob/810971de2b/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/legacy/core/jproto/JKey.java#L37).
+> <!-- TODO verify this about TransactionContext in system contracts -->
+
+## User stories
+
+<!-- Provide a list of “user stories” to express how this feature, functionality, improvement, or tool will be used by the end user. Template for user story: “As (user persona), I want (to perform this action) so that (I can accomplish this goal).” -->
+
+(1)
+As a smart contract developer,
+I want to code a smart contract in solidity which is able to identify whether a transaction is authorized when sent from an account with a complex key which contains a smart contract ID,
+so that I can implement advanced use cases including:
+
+- implement flexible and customisable authorisation rules
+- implement advanced multisig use cases without the need to split across multiple transactions
+- implement atomic multisig use cases
+- implement batch processing of multiple transactions
+
+(2)
+As a user of Hedera networks,
```

Comment:
Currently this is only possible - that I've been able to verify at least - with the HTS system contract,
being the only system contract that exposes non-`view` functions.

Whether or not this is possible depends on the implementation on HIP-632,
which is a pre-requisite for this HIP.

Also, even with HIP-632 implemented, this would still be blocked from happening,
because Solidity/ EVM op codes only expose a subset of transaction information,
which is insufficient to determine whether or not the current transaction is authorised.
This is the critical difference between `isAuthorized` and `isAuthorizedCurrentTransaction`.

Reply:
To clarify i'm saying it's possible for a contract to perform actions on behalf of an account if the contractId satisfies the account key requirement e.g. key is ContractID or 1 of 2 threshold with ContractId as 1 option as you've shown in some examples.
This 2nd User Story makes it seem like it's not possible and this HIP would make it possible.
Might need rewording or removal.